### PR TITLE
DCOS-13950: Fix sorting by health

### PR DESCRIPTION
**Pull request by weblancaster**
This PR fixes the issue when sorting by health in service table, adds minor refactoring (naming, code structure), unit test improvement based on code changes.

After talking to @mlunoe we decided that code fix was needed for the sorting to work properly as well.

As a effort to move from this [reusable sorting](https://github.com/dcos/dcos-ui/blob/2e63bbbc94875485c6dadfa9298649b42a6f213f/plugins/services/src/js/utils/TaskTableUtil.js#L19) because of reasons, code complexity, hard to follow code flow and hard to create more precise unit tests. 

That being said code refactoring was added, removing the sorting work from the `reusable sorting` function health and created a new method targeting specifically health sorting.

**Checklist**
- [x] Did you add a JIRA issue in a commit message or as part of the branch name?
- [x] Did you add new unit tests?
- [ ] Did you add new integration tests?
- [ ] If this is a regression, did you write a test to catch this in the future?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -17,3 +17,4 @@ jest/config.json
 jest/test-results
 webpack/proxy.dev.js
 src/resources/raml
+.nvmrc
```
Comment (weblancaster) [inline .gitignore]: Makes sense, will do next time if fine with everyone else.

Comment (mlunoe) [inline .gitignore]: Actually, let's remove it. I'll put up a PR to add it.
We don't want this in the `.gitignore`, if we want to add it to the project.

Comment (mlunoe) [inline .gitignore]: Created a PR here: https://github.com/dcos/dcos-ui/pull/2005

Comment (weblancaster) [inline .gitignore]: ok cool, thanks.

**Comment by Poltergeist** [comment]
How can I reproduce the initial issue? I am trying to check the PR, and cannot reproduce the issue.

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js [hunk 2]
```diff
@@ -9,13 +9,14 @@ describe('TaskTableUtil', function () {
   beforeEach(function () {
     this.getServiceHealth = MarathonStore.getServiceHealth;
     MarathonStore.getServiceHealth = function (prop) {
-      return this[prop].get('health');
+      return this[prop]['health'];
```
Comment (Poltergeist) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: The item should be of type `Item`. Which includes the get function so this line is fixing the issue but not the problem. Let's rather try to find the source of the problem. And ensure that all tasks are of type `Item`

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js [hunk 3]
```diff
@@ -52,7 +54,7 @@ describe('TaskTableUtil', function () {
       }
     });
 
-    this.sortFunction = TaskTableUtil.getSortFunction('name');
+    this.getComparator = TaskTableUtil.getSortFunction('name');
```
Comment (Poltergeist) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: If we decide to call the sort functions now comparator, which I do like, we should do this in a separate PR and apply it to the complete code base.

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js (hunk 2 above)
Comment (mlunoe) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: I don't see us using tasks as `Items` anywhere. We do this for Services.
Either we should make the effort and make everywhere we use tasks to structs or we keep this change to do as we do elsewhere when working with tasks.
This spawns a larger discussion about the future of structs in our code base and what the plan is to get where we want.
I am ok with this change as it follows what we do elsewhere.

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js (hunk 3 above)
Comment (mlunoe) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: I think we can still do this change in this PR. This helps readability of what is going on, but I do agree we should rename throughout the code base

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js (hunk 2 above)
Comment (weblancaster) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: @Poltergeist my main reason to remove the structs is that in this case is only creating more complexity in the code base because the get method is only a abstraction of the native notation which objects are being extended to use the method.

I'm totally fine changing to use the struct but I believe we want to move out using it.

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js (hunk 3 above)
Comment (weblancaster) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: I believe that changing incrementally as we go is a better option at the moment as developers work on files that needs to be refactor in a way to make more readable but I'm also fine removing from this PR and creating a Jira ticket to refactor the naming.

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js (hunk 2 above)
Comment (MatApple) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: @Poltergeist I agree with you that we shouldn't have inconsistencies. However, as @mlunoe points out - we've never enforced structs for tasks. In fact IMO, our structs (at least how they currently work via `Item.get('property')` causes far more complexity than was intended. We will be deprecating structs with Relay/GraphQL. Let's keep what @weblancaster has proposed as this will follow a concept that will persist (i.e. no structs).

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js (hunk 3 above)
Comment (MatApple) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: How about we leave this file as is AND create a JIRA ticket for renaming the comparator's? We should also be mindful that as we scale, it will be more difficult to enforce "soft" rules like this. We want to delegate as much as possible to the linter and outside of that we will have to pick our battles, keeping in mind that we also want to move fast and with community contribution. If something is easier to reason about let's use it and count it as an improvement.

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js (hunk 2 above)
Comment (Poltergeist) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: Okay I didn't know that we do not use structs on tasks my bad. I just saw the `item.get` and concluded that it was a struct. My bad then forget what I said here. Nice catch!

File: plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js (hunk 3 above)
Comment (Poltergeist) [inline plugins/services/src/js/utils/__tests__/TaskTableUtil-test.js]: 👍 MatApple, Let's keep the change and add the JIRA ticket.

File: plugins/services/src/js/utils/TaskTableUtil.js [hunk 4]
```diff
@@ -23,9 +23,9 @@ const TaskTableUtil = {
         return getUpdatedTimestamp(item) || 0;
       }
 
-      if (prop === 'health') {
+      if (prop === 'health' && !hasGetter ) {
```
Comment (mlunoe) [inline plugins/services/src/js/utils/TaskTableUtil.js]: Tiny #nit: You have an extra space inside the if-statement at the end

Comment (weblancaster) [inline plugins/services/src/js/utils/TaskTableUtil.js]: good catch

**Comment by mlunoe** [comment]
I might be missing something, but I think it is not working as expected:

**Comment by weblancaster** [comment]
Just fyi I have been working on a change for this PR based on my conversation with @mlunoe I'm also fixing the sorting besides the error thrown (the current changes here are only because I changed computers so I pushed what I had).

**Comment by weblancaster** [comment]
**before:**
https://cl.ly/3s0x430O0z1Y/Screen%20Recording%202017-03-22%20at%2005.14%20PM.gif

**after:**
https://cl.ly/0I1M2k2Z2A33/download/Screen%20Recording%202017-03-22%20at%2005.24%20PM.gif

**Comment by orlandohohmeier** [comment]
For you next PR, please don't squash all changes into one commit, so that we have a complete log of the things changed and the reasoning behind those changes.

**Comment by weblancaster** [comment]
Sounds good @orlandohohmeier I will start doing micro-commits going forward for big PRs changes, but you still get a complete log view of the changes looking the files changed and the reasons behind should be on the description.

**Comment by weblancaster** [comment]
retest this please

**Comment by wavesoft** [comment]
Looks, works great! Good work ✨  👍

**Comment by mlunoe** [comment]
🎉 Yaaaay!
